### PR TITLE
integrated chromaDB workflow with notebook examples

### PR DESCRIPTION
**Pull request by angel-daza**
ChromnaDB working with code as unintrusive as possible. The examples are in 3 separate notebooks and the update includes some few changes to the `Model` and `Corpus` classes.

### CODE REVIEW

File: tempo_embeddings/embeddings/vector_database.py [hunk 1]
```diff
@@ -0,0 +1,336 @@
...
+
+
+class ChromaDatabaseManager(VectorDatabaseManagerWrapper):
+    """A Chroma Database can create N collections and will always use THE SAME embedder function and tokenizer for all collections.
+        To create collections with different embedders one needs to create separate databases
+    """
+    def __init__(self, 
+                 db_path: str = "default_db",
+                 embedder_name: str = "sentence-transformers/all-MiniLM-L6-v2",
+                 embedder_config: Dict[str,Any] = None,
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:76]:
```suggestion
                 embedder_config: dict[str,Any] = None,
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 2]
```diff
@@ -0,0 +1,336 @@
...
+        if self.embedder_config and self.embedder_config.get("type") == "hf":
+            self.embedding_function = embedding_functions.HuggingFaceEmbeddingFunction(
+                api_key=self.embedder_config.get("api_key"),
+                model_name=embedder_name
+            )
+        elif self.embedder_config and self.embedder_config.get("type") == "custom":
+            self.requires_explicit_embeddings = True
+            self.embedding_function = None
+        else:
+            self.embedding_function = None
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:94]: 1. The `self.embedder_config.get("type")` call makes the `if self.embedder_config` condition redundant, because it returns `None` by default.
2. This logic seems to silently ignore cases where an invalid/unrecognized value for `self.embedder_config` has been given, which could result in unexpected behaviour in case of user typos etc.

File: tempo_embeddings/embeddings/vector_database.py [hunk 3]
```diff
@@ -0,0 +1,336 @@
...
+            self.client = chromadb.PersistentClient(path=self.db_path)
+            # If the given path existed then we will load the records AND override the parameters with the config
+            config_path = f"{self.db_path}/config.json"
+            if os.path.exists(config_path):
+                self.config = json.load(open(config_path))
+                print(f"Database Path '{self.db_path}' already exists. Loading current configuration\n{self.config}")
+            else:
+                print(f"Creating NEW Database in {self.db_path}...")
+                self._save_config()
+        else:
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:118]: I try to avoid `if not`-`else` clauses, so I would suggest to invert this to:
```python
if self.client:
  ...
else:
  ...
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+    def _save_config(self):
+        if self.config:
+            json.dump(self.config, open(f"{self.db_path}/config.json", "w"), indent=2)
+
+    def connect(self):
+        if not self.client:
+            self.client = chromadb.PersistentClient(path=self.db_path)
+            # If the given path existed then we will load the records AND override the parameters with the config
+            config_path = f"{self.db_path}/config.json"
+            if os.path.exists(config_path):
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:112]: Does that mean that we ignore the configuration that was passed to the `__init__()` method? If so, I think there should at least be a warning, perhaps even an exception.

File: tempo_embeddings/embeddings/vector_database.py [hunk 5]
```diff
@@ -0,0 +1,336 @@
...
+                self.config = json.load(open(config_path))
+                print(f"Database Path '{self.db_path}' already exists. Loading current configuration\n{self.config}")
+            else:
+                print(f"Creating NEW Database in {self.db_path}...")
+                self._save_config()
+        else:
+            print(f"A connection to the client already exsists in this session: {self.client}")
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:119]:
```suggestion
            logging.info(f"A connection to the client already exists in this session: {self.client}")
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 6]
```diff
@@ -0,0 +1,336 @@
...
+
+    def create_new_collection(self, name: str, passages: List[Passage] = None, embeddings: ArrayLike = None) -> Optional[Collection]:
+        if not self.client:
+            print("Please connect to a valid database first")
+            return None
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:124]: I think I would raise an exception here because this should never happen intentionally.

```suggestion
            raise RuntimeError("Please connect to a valid database first")
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 7]
```diff
@@ -0,0 +1,336 @@
...
+        # Create NEW collection and Embeds the given passages. Do nothing otherwise
+        try:
+            if self.embedding_function:
+                collection = self.client.create_collection(name, 
+                                                            embedding_function=self.embedding_function,
+                                                            metadata={"hnsw:space": "cosine"})
+            else:
+                collection = self.client.create_collection(name, metadata={"hnsw:space": "cosine"})
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:132]: Honestly, I find this part a bit hard to parse. Is the `if` block necessary at all, or could we just pass `embedding_function=None`? In that case, we can just do the same in both cases:
```suggestion
            collection = self.client.create_collection(name, 
                                                            embedding_function=self.embedding_function,
                                                            metadata={"hnsw:space": "cosine"})
```
If not, I would suggest storing the default value for `metadata` in a class field so that it can be updated in a single place in the future.

File: tempo_embeddings/embeddings/vector_database.py [hunk 8]
```diff
@@ -0,0 +1,336 @@
...
+            self.config["existing_collections"].append(name)
+            self.active_collection_name = name
+            self._save_config()
+            # If the collection is new then insert the corresponding passages already
+            if passages: self.insert_passages_embeddings(collection, passages, embeddings)
+
+            return collection
+        except UniqueConstraintError:
+            print(f"Collection '{name}' already exists. Not doing anything. (You can delete it or use the 'get_existing_collection' method)")
+            return None
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:142]: Why do we need to catch this exception at all? Is there a use case in which we would like to run this with an existing collection name and silently continue?

File: tempo_embeddings/embeddings/vector_database.py [hunk 9]
```diff
@@ -0,0 +1,336 @@
...
+            self.active_collection_name = name
+            return collection
+        except:
+           return None
+
+
+    def delete_collection(self, name):
+        try:
+            self.client.delete_collection(name)
+            print(f"Succesfully deleted {name}")
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:156]:
```suggestion
            logging.info(f"Succesfully deleted {name}")
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 10]
```diff
@@ -0,0 +1,336 @@
...
+        except:
+            print(f"Collection '{name}' does not exist in this database.")
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:158]: Catching all exceptions is quite risky because this implicitly handles exceptions that you might not have expected here at all. In this case, I would tend to not catching any exception, hence pass them on to the user (unless we know specifically which exceptions to expect and how to handle them).

File: tempo_embeddings/embeddings/vector_database.py [hunk 11]
```diff
@@ -0,0 +1,336 @@
...
+    def insert_passages_embeddings(self, collection: Collection, passages: List[Passage], embeddings: ArrayLike = None):
+        if self.requires_explicit_embeddings and embeddings is None:
+            raise ValueError("This Database requires embeddings to be explicitly pre-computed and fed into this function!")
+        
+        if isinstance(embeddings, list):
+            embeddings_list = embeddings
+        else:
+            embeddings_list = embeddings.tolist()
+        
+        # TODO: When executed in separate context it actually inserts records again (so Hash is not actually unique). Find a better UNIQUE ID!
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:170]:
```suggestion
        # TODO: When executed in separate context it actually inserts records again (so Hash is not actually unique). Find a better UNIQUE ID! See https://github.com/Semantics-of-Sustainability/tempo-embeddings/issues/40
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 12]
```diff
@@ -0,0 +1,336 @@
+import os, json
+import shutil
+from abc import ABC, abstractmethod
+import numpy as np
+from tqdm import tqdm
+from typing import Iterable, List, Callable, Dict, Any, Optional
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:6]: As above, no need to rely on `List` and `Dict` here.
```suggestion
from typing import Iterable, Callable, Any, Optional
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 13]
```diff
@@ -0,0 +1,336 @@
...
+          
+    def _tokenize(self, sentence: str):
+        encoded_input = self.tokenizer(sentence, return_tensors='pt')
+        return encoded_input
+
+    def embed_text_batch(self, text_batch: List[str]):
+        if self.embedding_function:
+            return self.embedding_function(text_batch)
+        else:
+            print("WARN: There is no valid embedding function in this database")
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:297]:
```suggestion
            logging.warning("There is no valid embedding function in this database")
```

**Review by carschno — CHANGES_REQUESTED** [review]
Hi,
The code looks good, thanks! Before merging, a few things need to be fixed:

1. The `chromadb` dependency needs to be added
1. Style/linting issues

The second point can be reproduced locally by running `prospector`. I use [ruff](https://github.com/astral-sh/ruff) for auto-formatting. At some point, I would even like to re-setup the project to replace all the other linting tests with `ruff`...

I have added a few comments and suggestions wrt style. This is of course not particularly important, but perhaps good to follow these patterns for consistency:
1. Do not catch exceptions (silently) unless they are clearly expected and handled properly.
1. Instead of `print()` calls, always use `logging.info()`/`warning()`/`error()` so that users can apply custom logging configuration.

The first two items are indicated by the failing checks too. Feel free to ignore (or to fix/deactivate) the `markdown-link-check` check.

File: tempo_embeddings/embeddings/vector_database.py (hunk 8 above)
Comment (angel-daza) [reply tempo_embeddings/embeddings/vector_database.py:142]: yes, currently there is... but perhaps is nicer to discuss the whole workflow in the notebook example so you tell me your opinions about how we can shape the code. The fact that it is a notebook confuses me about how defensive the code should be

File: setup.cfg [hunk 14]
```diff
@@ -50,6 +50,8 @@ install_requires =
     bokeh~=3.3.2
     holoviews
     dash~=2.11.1
+    # Required for Vector database:
```
Comment (carschno) [inline setup.cfg:53]: For conciseness, I prefer to have in-line comments only to resolve potentially misleading lines. This one seems pretty clear, no need to explain.
```suggestion
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 15]
```diff
@@ -1,15 +1,14 @@
...
+import os
+import json
+import logging
 from abc import ABC, abstractmethod
 import numpy as np
 from tqdm import tqdm
-from typing import Iterable, Callable, Any, Optional
-from typing import TYPE_CHECKING
+from typing import Iterable, Any, Optional
+# from typing import TYPE_CHECKING
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:8]: If the import is not needed any longer, it can be removed.
```suggestion
```

File: tempo_embeddings/embeddings/vector_database.py [hunk 16]
```diff
@@ -146,19 +146,21 @@ def get_existing_collection(self, name: str) -> Optional[Collection]:
             collection = self.client.get_collection(name, embedding_function=self.embedding_function)
             self.active_collection_name = name
             return collection
-        except:
+        except Exception as e:
```
Comment (carschno) [inline tempo_embeddings/embeddings/vector_database.py:148]: Minor remark: catching the generic `Exception` is equivalent to just running `except`. If possible, you would only catch the specific exceptions that can be expected here.

**Review by carschno — APPROVED** [review]
Looks good!
I only added some minor, non-mandatory suggestions.

**Comment by carschno** [comment]
I have noticed only now that the [linter has failed](https://github.com/Semantics-of-Sustainability/tempo-embeddings/actions/runs/7728846376/job/21070637829?pr=41). These should be fixed before merging.

**Comment by angel-daza** [comment]
yes, that one can be closed as well!
